src/cdomains/intDomain.ml

```diff
@@ -349,6 +349,446 @@ struct
+        if list_merged = [] then None else Some (SS.of_list list_merged)
+      | x1 :: xs1, x2 :: xs2, _ ->
+        let (y1, y2) = f x1 x2 in
+        if Int64.compare y1 min_int = 0 && Int64.compare y2 max_int = 0 then top ()
+          else arithm ~acc1:(x1 :: acc1) ~acc2:((y1,y2) :: acc2) f xs1 l2
+      | [], x2 :: xs2, _ -> arithm ~acc1:[] ~acc2:acc2 f acc1 xs2
+      | [], [], [] | _ :: _, [], [] -> failwith "error: empty entry"
+
+  let add_interval (x1, x2) (y1, y2) =
+    norm_interval ((Int64.add x1 y1), (Int64.add x2 y2))
+
+  let add x y =
+    match x, y with
+    | None, _ | _, None -> None
+    | Some a, Some b -> arithm add_interval (SS.elements a) (SS.elements b)
+
+  let neg_interval (x1, x2) _ =
+    norm_interval (Int64.neg x2, Int64.neg x1)
+
+  let neg x =
+    match x with
+    | None -> None
+    | Some s -> arithm neg_interval (SS.elements s) [(0L, 0L)]
+
+  let sub x y = add x (neg y)
+
+  let mul_interval (x1, x2) (y1, y2) =
+    let x1y1 = (Int64.mul x1 y1) in let x1y2 = (Int64.mul x1 y2) in
+    let x2y1 = (Int64.mul x2 y1) in let x2y2 = (Int64.mul x2 y2) in
+    norm_interval ((min (min x1y1 x1y2) (min x2y1 x2y2)),
+                   (max (max x1y1 x1y2) (max x2y1 x2y2)))
+
+  let mul x y =
+    match x, y with
+    | None, _ | _, None -> None
+    | Some s1, Some s2 -> arithm mul_interval (SS.elements s1) (SS.elements s2)
+
+  let rec div_interval (x1, x2) (y1, y2) =
+    begin match y1, y2 with
+      | 0L, 0L       -> (min_int, max_int)
+      | 0L, _        -> div_interval (x1,x2) (1L,y2)
+      | _      , 0L  -> div_interval (x1,x2) (y1,(-1L))
+      | _ when Interval32.leq (Some (0L, 0L)) (Some (y1,y2)) -> (min_int, max_int)
+      | _ ->
+        let x1y1n = (Int64.div x1 y1) in let x1y2n = (Int64.div x1 y2) in
+        let x2y1n = (Int64.div x2 y1) in let x2y2n = (Int64.div x2 y2) in
+        let x1y1p = (Int64.div x1 y1) in let x1y2p = (Int64.div x1 y2) in
+        let x2y1p = (Int64.div x2 y1) in let x2y2p = (Int64.div x2 y2) in
+        norm_interval ((min (min x1y1n x1y2n) (min x2y1n x2y2n)),
+                      (max (max x1y1p x1y2p) (max x2y1p x2y2p)))
+    end
+
+  let div x y =
+    match x, y with
+    | None, _ | _, None -> None
+    | Some s1, Some s2 -> arithm mul_interval (SS.elements s1) (SS.elements s2)
+
+  let join x y =
+    match x, y with
+    | None, z | z, None -> z
+    | Some s1, Some s2 ->
+      let l1 = SS.elements s1 in
+      let l2 = SS.elements s2 in
+      let l_joined = merge_intersect (List.sort IntPair.compare (l1 @ l2)) in
+      if l_joined = [] then None else Some (SS.of_list l_joined)
+
+  let meet x y =
+    match x, y with
+    | None, _ | _, None -> None
+    | Some s1, Some s2 -> meet_lists
+               (List.sort IntPair.compare (SS.elements s1))
+               (List.sort IntPair.compare (SS.elements s2))
+
+  (* the result of widening will always be one large interval *)
+  let widen x y =
```

Comment:
Maybe worth investigating if there is smarter things to do?
E.g. when you have `{[-2,4], [8,8]}` and you widen with `{[-2,-4], [8,9]}` you might want to go to `{[-2,-4], [8,\infty]}` instead of to `{[-2,\infty]}`.